contracts/ValidatorsRegistration.sol

```diff
@@ -0,0 +1,140 @@
+	IContractRegistry contractRegistry;
+
+	/*
+     * External methods
+     */
+
+    /// @dev Called by a participant who wishes to register as a validator
+	function registerValidator(bytes4 ip, address orbsAddr, string calldata name, string calldata website, string calldata contact) external {
+		require(!isRegistered(msg.sender), "Validator is already registered");
+		validators[msg.sender].registrationTime = now;
+		_updateValidator(ip, orbsAddr, name, website, contact);
+		emit ValidatorRegistered(msg.sender, ip, orbsAddr, name, website, contact);
+		// todo: notify elections contract?
+	}
+
+    /// @dev Called by a participant who wishes to update its propertires
+	function updateValidator(bytes4 ip, address orbsAddr, string calldata name, string calldata website, string calldata contact) external onlyRegisteredValidator {
+		_updateValidator(ip, orbsAddr, name, website, contact);
+		emit ValidatorDataUpdated(msg.sender, ip, orbsAddr, name, website, contact);
+		// todo: notify elections contract?
+	}
+
+    /// @dev Called by a prticipant to update additional validator metadata properties.
+    function setMetadata(string calldata key, string calldata value) external onlyRegisteredValidator {
+		string memory oldValue = validators[msg.sender].validatorMetadata[key];
+		validators[msg.sender].validatorMetadata[key] = value;
+		emit ValidatorMetadataChanged(msg.sender, key, value, oldValue);
+	}
+
+	function getMetadata(address addr, string calldata key) external view returns (string memory) {
+		require(isRegistered(addr), "Validator is not registered");
+		return validators[addr].validatorMetadata[key];
+	}
+
+	/// @dev Called by a participant who wishes to unregister
+	function unregisterValidator() external onlyRegisteredValidator {
+		orbsAddressToEthereumAddress[validators[msg.sender].orbsAddr] = address(0);
+		delete validators[msg.sender];
+		emit ValidatorUnregistered(msg.sender);
+		// todo: notify elections contract?
+	}
+
+    /// @dev Returns a validator's data
+    /// Used also by the Election contract
+	function getValidatorData(address addr) external view returns (bytes4 ip, address orbsAddr, string memory name, string memory website, string memory contact, uint registration_time, uint last_update_time) {
+		require(isRegistered(addr), "Validator is not registered");
+		Validator memory v = validators[addr];
+		return (v.ip, v.orbsAddr, v.name, v.website, v.contact, v.registrationTime, v.lastUpdateTime);
+	}
+
+	/*
+     * Methods restricted to other Orbs contracts
+     */
+
+    /// @dev Translates a list validators Ethereum addresses to Orbs addresses
+    /// Used by the Election conract
+	function getOrbsAddresses(address[] calldata ethereumAddrs) external view returns (address[] memory orbsAddrs) {
+		orbsAddrs = new address[](ethereumAddrs.length);
+		for (uint i = 0; i < ethereumAddrs.length; i++) {
+			require(isRegistered(ethereumAddrs[i]), "Validator is not registered"); // todo: can be optimized, or maybe omit?
```

Comment:
A require as part of the elections flow is risky and creates a double source of truth: a validator is registered if appears in committee / registered in registration contract. add a TODO to review